contracts/tokenbridge/ethereum/L1AtomicTokenBridgeCreator.sol

```diff
@@ -0,0 +1,454 @@
+    address public immutable canonicalL2FactoryAddress;
+    address public immutable canonicalL2ProxyAdminAddress;
+    address public immutable canonicalL2BeaconProxyFactoryAddress;
+
+    constructor() Ownable() {
+        canonicalL2FactoryAddress = _computeAddress(AddressAliasHelper.applyL1ToL2Alias(address(this)), 0);
+
+        canonicalL2ProxyAdminAddress = Create2.computeAddress(
+            _getL2Salt(OrbitSalts.L2_PROXY_ADMIN), keccak256(type(ProxyAdmin).creationCode), canonicalL2FactoryAddress
+        );
+
+        canonicalL2BeaconProxyFactoryAddress = Create2.computeAddress(
+            _getL2Salt(OrbitSalts.BEACON_PROXY_FACTORY),
+            keccak256(type(BeaconProxyFactory).creationCode),
+            canonicalL2FactoryAddress
+        );
+    }
+
+    /**
+     * @notice Set addresses of L1 logic contracts and L2 contracts which are deployed on L1.
+     * @dev L2 contracts are deployed to L1 as bytecode placeholders - that bytecode will be part of retryable
+     *      payload used to deploy contracts on L2 side.
+     */
+    function setTemplates(
+        L1GatewayRouter _router,
+        L1ERC20Gateway _standardGateway,
+        L1CustomGateway _customGateway,
+        L1WethGateway _wethGatewayTemplate,
+        address _l2TokenBridgeFactoryTemplate,
+        address _l2RouterTemplate,
+        address _l2StandardGatewayTemplate,
+        address _l2CustomGatewayTemplate,
+        address _l2WethGatewayTemplate,
+        address _l2WethTemplate,
+        address _l1Weth
+    ) external onlyOwner {
+        routerTemplate = _router;
+        standardGatewayTemplate = _standardGateway;
+        customGatewayTemplate = _customGateway;
+        wethGatewayTemplate = _wethGatewayTemplate;
+
+        l2TokenBridgeFactoryTemplate = _l2TokenBridgeFactoryTemplate;
+        l2RouterTemplate = _l2RouterTemplate;
+        l2StandardGatewayTemplate = _l2StandardGatewayTemplate;
+        l2CustomGatewayTemplate = _l2CustomGatewayTemplate;
+        l2WethGatewayTemplate = _l2WethGatewayTemplate;
+        l2WethTemplate = _l2WethTemplate;
+
+        l1Weth = _l1Weth;
+
+        emit OrbitTokenBridgeTemplatesUpdated();
+    }
+
+    /**
+     * @notice Deploy and initialize token bridge, both L1 and L2 sides, as part of a single TX.
+     * @dev This is a single entrypoint of L1 token bridge creator. Function deploys L1 side of token bridge and then uses
+     *      2 retryable tickets  to deploy L2 side. 1st one deploy L2 factory and 2nd calls function that deploys and inits
+     *      all the rest of the contracts. L2 chain is determined by `inbox` parameter.
+     */
+    function createTokenBridge(
+        address inbox,
+        uint256 maxSubmissionCostForFactory,
+        uint256 maxGasForFactory,
+        uint256 maxSubmissionCostForContracts,
+        uint256 maxGasForContracts,
+        uint256 gasPriceBid
+    ) external payable {
```

Comment:
maxSubmissionCost is dynamic (since it depends on L1 gas price), we can instead calculate it onchain and refund the excess msg.value to the sender.

Reply:
Also add a revert if this is called before setTemplates, tho it will prolly revert somewhere in L1 deploy.

Reply:
both done